waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,379 @@
+               "urn:oid:2.5.4.3" "cn"
+               "urn:oid:2.5.4.4" "sn"
+               "urn:oid:2.5.4.12" "title"
+               "urn:oid:2.5.4.20" "phone"
+               "urn:oid:2.5.4.42" "givenName"
+               "urn:oid:2.5.6.8" "organizationalRole"
+               "urn:oid:2.16.840.1.113730.3.1.3" "employeeNumber"
+               "urn:oid:2.16.840.1.113730.3.1.4" "employeeType"
+               "urn:oid:1.3.6.1.4.1.5923.1.1.1.1" "eduPersonAffiliation"
+               "urn:oid:1.3.6.1.4.1.5923.1.1.1.2" "eduPersonNickname"
+               "urn:oid:1.3.6.1.4.1.5923.1.1.1.6" "eduPersonPrincipalName"
+               "urn:oid:1.3.6.1.4.1.5923.1.1.1.9" "eduPersonScopedAffiliation"
+               "urn:oid:1.3.6.1.4.1.5923.1.1.1.10" "eduPersonTargetedID"
+               "urn:oid:1.3.6.1.4.1.5923.1.6.1.1" "eduCourseOffering"}]
+    (fn [attr-oid]
+      (get names attr-oid attr-oid))))
+
+;; http://kevnls.blogspot.gr/2009/07/processing-saml-in-java-using-opensaml.html
+;; http://stackoverflow.com/questions/9422545/decrypting-encrypted-assertion-using-saml-2-0-in-java-using-opensaml
+(defn parse-saml-assertion
+  "Returns the attributes and the 'audiences' for the given SAML assertion"
+  [assertion]
+  (let [statements (.getAttributeStatements assertion)
+        subject (.getSubject assertion)
+        subject-confirmation-data (.getSubjectConfirmationData
+                                    (first (.getSubjectConfirmations subject)))
+        name-id (.getNameID subject)
+        attributes (mapcat #(.getAttributes %) statements)
+        attrs (apply merge
+                     (map (fn [a] {(saml2-attr->name (.getName a)) ;; Or (.getFriendlyName a) ??
+                                   (map #(-> % (.getDOM) (.getTextContent))
+                                        (.getAttributeValues a))})
+                          attributes))
+        conditions (.getConditions assertion)
+        audiences (mapcat #(let [audiences (.getAudiences %)]
+                             (map (fn [a] (.getAudienceURI a)) audiences))
+                          (.getAudienceRestrictions conditions))]
+    {:attrs attrs :audiences audiences
+     :name-id
+     {:value (when name-id (.getValue name-id))
+      :format (when name-id (.getFormat name-id))}
+     :confirmation
+     {:in-response-to (.getInResponseTo subject-confirmation-data)
+      :not-before (tc/to-timestamp (.getNotBefore subject-confirmation-data))
+      :not-on-or-after (tc/to-timestamp (.getNotOnOrAfter subject-confirmation-data))
+      :recipient (.getRecipient subject-confirmation-data)}}))
+
+(let [authenticated-redirect-template-fn
+      (template/fn
+        [{:keys [auth-redirect-uri saml-auth-data]}]
+        (slurp (io/resource "web/authenticated-redirect.html")))]
+  (defn render-authenticated-redirect-template
+    "Render the authenticated redirect html page."
+    [context]
+    (authenticated-redirect-template-fn context)))
+
+(defn saml-acs-handler
```

Comment:
We ran into performance issues with the kerberos auth handling. We should introduce metrics for time spent on these computations. Eventually, we might need to introduces throttling via a custom thread pool like we did for kerberos.

Reply:
done